### PR TITLE
fix has_aux bug in gauss_newton

### PR DESCRIPTION
**Pull request by amir-saadat**
@mblondel @vroulet , 
Making some improvements and bug fixes for GaussNewton. In particular, I'm making some minor changes in the printing error section so that we can Jit functions while using verbose=True. This should also let other solvers be jittable while using verbose=True.

### CODE REVIEW

**Comment by vroulet** [comment]
Hello Amir, 
Sorry for the delay I wanted the previous PR on LM to be done. 
Concerning this one:
- Using jax.debug for the prints is great. It would probably be best to have this done outside of this PR and review all algorithms. For example, one may simplify the printing of the errors in the zoom linesearch too and revise all algorithms that give additional information when verbose=True. I can do that if you don't have the bandwidth.
- For Gauss-Newton, I wasn't there when it was first implemented and I would have some generic questions before enhancing e.g. the jit possibilities:
  - Gauss-Newton *needs* a linesearch to converge otherwise it is well-known that it can diverge (see Wikipedia counter-example for example). 
  - I don't really see any use-case for a GN method if we have a LM method that is functional. True Gauss-Newton has a quadratic local convergence but globally it can be pretty bad. On the other hand, LM can better behave globally. True LM with fixed regularization parameter only converges linearly but with varying regularization, I believe the quadratic convergence can be ensured. Overall I think LM is more general anyway. What do you think?

**Comment by vroulet** [comment]
> Sounds good, I moved it to https://github.com/google/jaxopt/pull/503

Thanks, that'll be a very useful addition!

> This would then become an LM variant right?

Not exactly, LM uses a trust-region method. I was saying a simple linesearch. By the way, it would be interesting to have a generic trust-region method which could apply not only to LM.

> I agree with you that LM would be preferred than GN ...

Ok! So shall we close this PR then for now if that's ok with you?

File: jaxopt/_src/gauss_newton.py [hunk 1]
```diff
@@ -119,6 +120,7 @@ def update(self,
     state = GaussNewtonState(iter_num=state.iter_num + 1,
                              error=tree_l2_norm(delta_params),
                              residual=residual,
+                             loss=0.5 * (residual @ residual),
```
Comment (vroulet) [inline jaxopt/_src/gauss_newton.py:123]: loss -> value (that's the current name given in other solvers).

File: jaxopt/_src/gauss_newton.py [hunk 2]
```diff
@@ -91,6 +92,7 @@ def init_state(self,
     return GaussNewtonState(iter_num=jnp.asarray(0),
                             error=jnp.asarray(jnp.inf),
                             residual=residual,
+                            loss=0.5 * (residual @ residual),
```
Comment (vroulet) [inline jaxopt/_src/gauss_newton.py:95]: loss -> value

File: jaxopt/_src/gauss_newton.py [hunk 3]
```diff
@@ -33,6 +33,7 @@ class GaussNewtonState(NamedTuple):
   """Named tuple containing state information."""
   iter_num: int
   residual: Any
+  loss: Any
```
Comment (vroulet) [inline jaxopt/_src/gauss_newton.py:36]: loss -> value

File: jaxopt/_src/gauss_newton.py (hunk 1 above)
Comment (vroulet) [inline jaxopt/_src/gauss_newton.py:123]: As you're on it, make it able to handle scalars as you did for LM.

File: tests/gauss_newton_test.py [hunk 4]
```diff
@@ -69,6 +69,21 @@ def setUp(self):
     ])
     self.init_temperature_record_coeffs = onp.array([10, 0.5, 10.5, 50])
 
+  def test_aux_true(self):
```
Comment (vroulet) [inline tests/gauss_newton_test.py:72]: Make it a one step verification (like max_iter =2) with a fixed aux for any state (like True) and check if the aux is still the same. I think other solverrs have such tests. The current tests are already taking a long time so it's preferable to minimize the lengths of the tests if we can.

File: tests/gauss_newton_test.py [hunk 5]
```diff
@@ -80,8 +95,8 @@ def test_enzyme_reaction_parameter_fit(self):
         self.rate_data)
 
     self.assertArraysAllClose(optimize_info.params,
-                              onp.array([0.36183689, 0.55626653]))
-
+                              onp.array([0.36183689, 0.55626653]),
```
Comment (vroulet) [inline tests/gauss_newton_test.py:98]: Where does this value come from? 
- if you have a ref add it
- otherwise, 
  - you may simply compare gn to a solver like LBFGS
  - you may create artifically some nonlinear data, add some noise then compare GN to the solution of LBFGS initialized at the parameter that generated the data.
  
Note that this is optional, that is, since we may redo this class later, we can pass that. But if you have a ref this will be helpful later.

File: tests/gauss_newton_test.py [hunk 6]
```diff
@@ -115,8 +130,8 @@ def test_temperature_record_four_parameter_fit(self):
     # Checking against the expected values
     self.assertArraysAllClose(
         optimize_info.params,
-        onp.array([16.63994555, 0.46327812, 10.85228919, 76.19086103]))
-    self.assertAllClose(optimize_info.state.iter_num, 30)
+        onp.array([16.63994555, 0.46327812, 10.85228919, 76.19086103]),
```
Comment (vroulet) [inline tests/gauss_newton_test.py:133]: Same as above, a ref would be good if you have one.

**Comment by vroulet** [comment]
Thanks for the PR! I left you some comments.

**Comment by vroulet** [comment]
Hello Amir, 
If the comments above were fixed can you mark them as resolved and squash everything? 
I'll add the pull ready tag.
Thanks for all!
